recipes/rapidgzip/THIRDPARTY.yml

```diff
@@ -0,0 +1,27 @@
+---
```

Comment:
What's the origin of this file?

Reply:
`rapidgzip --oss-attributions`

Reply:
- cxxopts: is a header-only library, dynamic linking makes no sense
 - ISA-L a custom fork is used and upstream cannot be used because of custom additions
 - rpmalloc: I did not bother dynamically linking against this, because it is not available as a package on Ubuntu. It can be disabled with the `WITH_RPMALLOC` CMake option, which seems like the easiest option if static linking really is that much disregarded.
 - zlib: can be dynamically linked with the `USE_SYSTEM_ZLIB` CMake option. I don't know how to set this option for the conda build. I guess I'll need to take a look.

Reply:
If zlib is dynamically linked, is that a reason to remove if from THIRDPARTY.yml?

Reply:
If this file can be generated, it should be generated during the build and not be submitting as part of the recipe.

Reply:
I have finally tried to use cxxopts and zlib via the existing conda packages but I get an import error because of an undefined symbol now:

    The following NEW packages will be INSTALLED:

        _libgcc_mutex:    0.1-conda_forge            conda-forge
        _openmp_mutex:    4.5-2_gnu                  conda-forge
        bzip2:            1.0.8-hd590300_5           conda-forge
        ca-certificates:  2023.11.17-hbcca054_0      conda-forge
        cxxopts:          3.1.0-hf52228f_0           conda-forge
        ld_impl_linux-64: 2.40-h41732ed_0            conda-forge
        libffi:           3.4.2-h7f98852_5           conda-forge
        libgcc-ng:        13.2.0-h807b86a_3          conda-forge
        libgomp:          13.2.0-h807b86a_3          conda-forge
        libnsl:           2.0.1-hd590300_0           conda-forge
        libsqlite:        3.44.2-h2797004_0          conda-forge
        libstdcxx-ng:     13.2.0-h7e041cc_3          conda-forge
        libuuid:          2.38.1-h0b41bf4_0          conda-forge
        libzlib:          1.2.13-hd590300_5          conda-forge
        ncurses:          6.4-h59595ed_2             conda-forge
        openssl:          3.2.0-hd590300_1           conda-forge
        pip:              23.3.1-pyhd8ed1ab_0        conda-forge
        python:           3.10.13-hd12c33a_0_cpython conda-forge
        python_abi:       3.10-4_cp310               conda-forge
        rapidgzip:        0.10.4-py310hd5d28ae_0     local      
        readline:         8.2-h8228510_1             conda-forge
        setuptools:       68.2.2-pyhd8ed1ab_0        conda-forge
        tk:               8.6.13-noxft_h4845f30_101  conda-forge
        tzdata:           2023c-h71feb2d_0           conda-forge
        wheel:            0.42.0-pyhd8ed1ab_0        conda-forge
        xz:               5.2.6-h166bdaf_0           conda-forge
        zlib:             1.2.13-hd590300_5          conda-forge

    Preparing transaction: ...working... done
    Verifying transaction: ...working... done
    Executing transaction: ...working... done
    export PREFIX=/miniconda3/conda-bld/rapidgzip_1701540286573/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place
    export SRC_DIR=/miniconda3/conda-bld/rapidgzip_1701540286573/test_tmp
    Traceback (most recent call last):
      File "/miniconda3/conda-bld/rapidgzip_1701540286573/test_tmp/run_test.py", line 2, in <module>
    import: 'rapidgzip'
        import rapidgzip
    ImportError: /miniconda3/conda-bld/rapidgzip_1701540286573/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/site-packages/rapidgzip.cpython-310-x86_64-linux-gnu.so: undefined symbol: inflatePrime
    WARNING: Tests failed for rapidgzip-0.10.4-py310hd5d28ae_0.tar.bz2 - moving package to /miniconda3/conda-bld/broken
    TESTS FAILED: rapidgzip-0.10.4-py310hd5d28ae_0.tar.bz2

According to the [changelog](https://github.com/madler/zlib/blob/develop/ChangeLog), `inflatePrime` has been added in 2005 with version 1.2.2.4 and according to the list above, version 1.2.13 is loaded. So, why is `inflatePrime` still missing? And it seems to be the only missing symbol as many others are used but not shown as missing.

I have attached the patch, which probably should be cherry-picked by @ap-- ontop of this PR if only it would work ...

[0001-Use-conda-packaged-cxxopts-and-zlib.txt](https://github.com/conda-forge/staged-recipes/files/13537031/0001-Use-conda-packaged-cxxopts-and-zlib.txt)

Reply:
> cxxopts: is a header-only library, dynamic linking makes no sense. I don't understand what the conda-forge package exists for. I guess to update the headers and rebuild dependant conda-forge packages?

Not every library revends its dependencies, so the package is there for convenience. If CXXOPTS is header only, then static linking is the only option, so that's the justification.

> ISA-L: uses a custom fork made by me. Upstream cannot be used because of custom additions. It can be disabled with the WITH_ISAL CMake option. Rapidgzip is ~2x faster with ISA-L. (The upstream version would have been available on conda-forge)

If your custom fork contains unique changes that are not compatible with upstream, this is an acceptable justification because your fork is no longer the canonical ISA-L. However, we encourage you to upstream your changes, so others can benefit.

> rpmalloc: I did not bother dynamically linking against this, because it is not available as a package on Ubuntu. It can be disabled with the WITH_RPMALLOC CMake option, which seems like the easiest option. Rpmalloc only leads to ~20% speedup for short workloads and possibly for longer workloads on high core counts. It is much less important for performance than ISA-L. I did not find rpmalloc on conda-forge.

Create an additional recipe folder for rpmalloc in this PR, and add rpmalloc as a host dependency for rapidgzip. Our build runner can determine in which order to build multiple recipes or disable this option until rpmalloc is added.

> zlib: can be dynamically linked with the USE_SYSTEM_ZLIB CMake option. I don't know how to set this option for the conda build. I guess I'll need to take a look. Unsurprisingly exists on conda-forge.

Add zlib to the host requirements. Set USE_SYSTEM_ZLIB and ensure that your package links to libz.so in the host environment i.e. $PREFIX during the build step. Please push the commits in which you attempt to USE_SYSYEM_ZLIB so we can read the build logs and/or recreate your work.